### DISCUSSION TITLE
Urgent CSRF issue

### DISCUSSION THREAD

**Discussion by risenhoover** [playframework/playframework#11364] [Q&A]
Play Framework 2.7.9 -> As of three days ago, we started getting reports of people not being able to login, even if they present valid credentials. Upon investigation, they were getting the gray "unauthorized" page provided by the Framework. This happens with various people but others are able to login without issues. We use SecureSocial for authentication and authorization.

The log reports the following message: "[warn] - play.filters.CSRF - [CSRF] Check failed because no token found in headers for /auth/authenticate/userpass"

Additionally, if we open a private window, we can reproduce this error ad infinitum - it's never possible to login and get past the CSRF check and failure. We are also able to see this same issue on other forms protected by CSRF token, in particular our payment page :(.

I have verified that the login page does in fact have a CSRF token in the form, and it is being submitted properly. What is strange is that if you look at the message and the source ( https://github.com/playframework/playframework/blob/d8ebf7cbdb8d360ee499333cc853de708a9eac49/web/play-filters-helpers/src/main/scala/play/filters/csrf/CSRFActions.scala )

it says "NO TOKEN FOUND" - so somewhere somehow this token is getting lost in transit.

**Comment by Squeng** [comment]
I'm using Play 2.8 and I don't use SecureSocial so I won't be of much help, but I can provide a few sanity checks:

- If you apply "+ nocsrf" to the login route, does it work again? https://www.playframework.com/documentation/2.7.x/ScalaCsrf#Applying-a-global-CSRF-filter
- If you look at the PLAY_SESSION cookie (or the CSRF cookie, if you have a dedicated one https://www.playframework.com/documentation/2.7.x/ScalaCsrf#CSRF-configuration-options), does it contain the CSRF token? (You may have to decode it with https://jwt.io/)
![grafik](https://user-images.githubusercontent.com/7399173/180372308-1d7b9f26-57e6-48cb-be53-e9cc00bac9e4.png)


Best of luck!
Paul

**Comment by mkurz** [comment]
@risenhoover To be honest, since we don't support Play 2.7.x anymore, I highly recommend to upgrade to latest Play 2.8.x, as our resources are limited right now and we don't have time to look into the 2.7.x code base anymore. As far as I remember there have been CSRF fixes that made it into 2.8.x, so your problem might already be fixed.
However if you can provide a reproducer project that uses latest Play 2.8.x we could take a look what's going on. Since you are using secure social, that might also cause the problem.

**Comment by brainandbinary** [comment]
This happens for many reasons
1. if you use **session domain** in config file but at browser the **url domain** is different one , so the origin of cookie mismatch then the problem happens

2. if you pragmatically delete the session by using `withNewSession` method, this can happen. In this case explicitly setting the csrf-token is recommended from me. 
```
def logout: Action[AnyContent] = Action {
    implicit request =>
      val csrfToken = CSRF.getToken.get.value
      
      Ok(logIn("Please log in")).withNewSession.withSession("csrfToken"->csrfToken)

  }  
```
3. Another reason can be you changed your application secret key, but your user are using old secret key certified session

Basically for these reasons this problem happens , I saw.
